### PR TITLE
Change operator api to support dual-stack enablement of Istio components

### PR DESCRIPTION
**Pull request by jacob-delgado**
**Please provide a description of this PR:**

Help users setup Istio components (istiod, gateways) and be able to set dual-stack environment variables to enable Istio to work in dual-stack Kubernetes clusters.

Partial fix for https://github.com/istio/istio/issues/44017

### CODE REVIEW

**Comment by jacob-delgado** [comment]
Adding a DNM/hold until we get consensus on the approach. There is still some testing for me to do, but I want to make sure we are fine with these changes.

File: operator/pkg/apis/istio/v1alpha1/values_types.proto [hunk 1]
```diff
@@ -1230,3 +1241,8 @@ message IntOrString {
 
   google.protobuf.StringValue strVal = 3;
 }
+
+message DualStack {
+   // Controls whether CNI is enabled.
```
Comment (kfaseela) [inline operator/pkg/apis/istio/v1alpha1/values_types.proto:1246]:
```suggestion
   // Controls whether dual-stack is enabled.
```

File: operator/pkg/apis/istio/v1alpha1/values_types.proto [hunk 2]
```diff
@@ -1024,13 +1031,17 @@ message SecretVolume {
...
 // ServiceConfig is described in istio.io documentation.
 message ServiceConfig {
-  google.protobuf.Struct annotations = 1;
-
-  uint32 externalPort = 2;
-
-  string name = 3;
-
-  string type = 18;
+  repeated string ipFamilies = 4;
```
Comment (kfaseela) [inline operator/pkg/apis/istio/v1alpha1/values_types.proto:1034]: These can go to the end?

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml [hunk 3]
```diff
@@ -136,6 +136,10 @@ spec:
     - name: TRUST_DOMAIN
       value: "{{ . }}"
     {{- end }}
+    {{- if .Values.global.dualStack.enabled }}
+    - name: ISTIO_DUAL_STACK
```
[the remark this answers is not in the document]
Comment (jacob-delgado) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:140]: I'll remove it. However, I'm not sure if we will ever get it to be zero config. There's really no way to tell if you are running in a dual-stack environment. Kubernetes has a difficult time itself distinguishing if you want it to be enabled and there is no way to query the apiserver or kubelet to determine if it was set. The only approach I've thought of at this time is to get the endpoints and see if you have a mix of IPv4 or IPv6 endpoints. However, that's only true if you've deployed in a dual-stack cluster and you have migrated at least 1 service. Not exactly ideal for people who just enabled it. I don't think the path forward for someone enabling dual-stack in Kubernetes is a rather kind road which is why many people haven't adopted it yet.

[the remark this answers is not in the document]
Comment (keithmattix) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:140]: There's some functionality associated with ISTIO_DUAL_STACK; DNS lookup family (opt into happy eyeballs) is one that I found on a quick lookthrough of the code

File: operator/pkg/apis/istio/v1alpha1/values_types.proto (hunk 2 above)
Comment (jacob-delgado) [reply operator/pkg/apis/istio/v1alpha1/values_types.proto:1034]: This message isn't being Used. I just removed it.

File: manifests/charts/gateway/templates/service.yaml [hunk 4]
```diff
@@ -15,12 +15,11 @@ spec:
 {{- with .Values.service.loadBalancerIP }}
   loadBalancerIP: "{{ . }}"
 {{- end }}
-{{- with .Values.service.ipFamilyPolicy }}
-  ipFamilyPolicy: "{{ . }}"
+{{- if ne .Values.service.serviceConfig.ipFamilyPolicy "" }}
```
[the remark this answers is not in the document]
Comment (jacob-delgado) [reply manifests/charts/gateway/templates/service.yaml:18]: Removed it.

File: manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacob-delgado) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:140]: I removed that functionality for now to simplify this PR. There's still plenty of work for us to do to get dual-stack enabled. It's fine if users need to set the environment variables themselves.

File: operator/pkg/apis/istio/v1alpha1/values_types.proto [hunk 5]
```diff
@@ -1022,17 +1033,6 @@ message SecretVolume {
   string secretName = 3;
 }
 
-// ServiceConfig is described in istio.io documentation.
-message ServiceConfig {
```
Comment (jacob-delgado) [inline operator/pkg/apis/istio/v1alpha1/values_types.proto:1026]: This is not being used by anything and hasn't been for a while.

**Comment by jacob-delgado** [comment]
Ready for another re-review @howardjohn and @kfaseela 

The following is passed though

```yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  components:
    base:
      enabled: true
    cni:
      enabled: true
    egressGateways:
    - enabled: true
      name: istio-egressgateway
    ingressGateways:
    - enabled: true
      name: istio-ingressgateway
    istiodRemote:
      enabled: true
    pilot:
      enabled: true
  hub: localhost:5000
  tag: istio-testing
  values:
    gateways:
      istio-egressgateway:
        ipFamilyPolicy: PreferDualStack
        ipFamilies: [IPv6, IPv4]
      istio-ingressgateway:
        ipFamilyPolicy: RequireDualStack
        ipFamilies: [IPv4, IPv6]
    global:
      configValidation: true
    pilot:
      ipFamilyPolicy: RequireDualStack
      ipFamilies: [IPv4, IPv6]
```

File: manifests/charts/base/templates/services.yaml [hunk 6]
```diff
@@ -24,5 +24,11 @@ spec:
   type: ExternalName
   externalName: {{ .Values.global.remotePilotAddress }}
   {{- end }}
+  {{- if .Values.global.ipFamilyPolicy }}
```
Comment (kfaseela) [inline manifests/charts/base/templates/services.yaml:27]: where do we use global.ipFamilyPolicy? Did not see that in the example you have used.

File: manifests/charts/base/templates/services.yaml [hunk 7]
```diff
@@ -24,5 +24,11 @@ spec:
...
+  ipFamilyPolicy: {{ .Values.global.ipFamilyPolicy }}
+  {{- end }}
+  {{- with .Values.global.ipFamilies }}
```
Comment (kfaseela) [inline manifests/charts/base/templates/services.yaml:30]: do we want to have an if check here as well?

File: manifests/charts/gateway/templates/service.yaml [hunk 8]
```diff
@@ -15,12 +15,14 @@ spec:
 {{- with .Values.service.loadBalancerIP }}
   loadBalancerIP: "{{ . }}"
 {{- end }}
-{{- with .Values.service.ipFamilyPolicy }}
-  ipFamilyPolicy: "{{ . }}"
+{{- if .Values.service.ipFamilyPolicy }}
+  ipFamilyPolicy: {{ .Values.pilot.ipFamilyPolicy }}
```
Comment (kfaseela) [inline manifests/charts/gateway/templates/service.yaml:19]: does this mean that we copy the value from the pilot config for gateway config, if not explicitly specified?

Comment (jacob-delgado) [reply manifests/charts/gateway/templates/service.yaml:19]: No, this is a bug. I'll fix it.

File: manifests/charts/base/templates/services.yaml (hunk 7 above)
Comment (jacob-delgado) [reply manifests/charts/base/templates/services.yaml:30]: I've added a comment that is largely intended for use with external istiod. I haven't ever installed external istiod (or even multicluster) with just the helm charts. Do you think I should remove? That may be another issue to track onto itself.

**Review by kfaseela — APPROVED** [review]

**Comment by jacob-delgado** [comment]
Deployed the following with istioctl

```yaml
kind: IstioOperator
spec:
  components:
    base:
      enabled: true
    cni:
      enabled: true
    egressGateways:
    - enabled: true
      name: istio-egressgateway
    ingressGateways:
    - enabled: true
      name: istio-ingressgateway
    istiodRemote:
      enabled: true
    pilot:
      enabled: true
  hub: localhost:5000
  tag: istio-testing
  meshConfig:
    defaultConfig:
      proxyMetadata:
        ISTIO_DUAL_STACK: "true"
  values:
    gateways:
      istio-egressgateway:
        ipFamilyPolicy: PreferDualStack
        ipFamilies: [IPv6, IPv4]
      istio-ingressgateway:
        ipFamilyPolicy: RequireDualStack
        ipFamilies: [IPv4, IPv6]
    global:
      configValidation: true
    pilot:
      env:
        ISTIO_DUAL_STACK: "true"
      ipFamilyPolicy: RequireDualStack
      ipFamilies: [IPv6, IPv4]
```

After it was deployed I verified that dual-stack services were deployed as well as endpointslices:

```bash
% k get svc -n istio-system                                                                                                                                                                                       ✹ ✭
NAME                   TYPE           CLUSTER-IP         EXTERNAL-IP                            PORT(S)                                      AGE
istio-egressgateway    ClusterIP      fd00:10:96::9642   <none>                                 80/TCP,443/TCP                               151m
istio-ingressgateway   LoadBalancer   10.96.72.234       172.17.7.110,fc00:f853:ccd:e793::76e   15021:30777/TCP,80:30354/TCP,443:30875/TCP   151m
istiod                 ClusterIP      fd00:10:96::4b27   <none>                                 15010/TCP,15012/TCP,443/TCP,15014/TCP        151m
(⎈|kind-istio-testing:default)[jacob@buildserver:~/go/src/istio.io/istio on service-proto-dual-stack]
% k get endpointslices.discovery.k8s.io -n istio-system                                                                                                                                                           ✹ ✭
NAME                         ADDRESSTYPE   PORTS                           ENDPOINTS        AGE
istio-egressgateway-9mpcd    IPv4          8080,8443                       10.244.0.9       151m
istio-egressgateway-rhpkl    IPv6          8080,8443                       fd00:10:244::9   151m
istio-ingressgateway-5gcpl   IPv6          8080,8443,15021                 fd00:10:244::8   151m
istio-ingressgateway-v7nps   IPv4          8080,8443,15021                 10.244.0.8       151m
istiod-92qf8                 IPv4          15012,15017,15010 + 1 more...   10.244.0.7       151m
istiod-zqvtb                 IPv6          15012,15017,15010 + 1 more...   fd00:10:244::7   151m
```

I deployed sleep and httpbin to the default namespace where sidecar injection worked properly.

I then port-forwarded to istiod using only IPv6 and verified that clients can connect to it using [pilot_cli](https://github.com/jacob-delgado/pilot_cli)

```bash
% kubectl port-forward --address "::1"  -n istio-system istiod-69dcbd84d4-zpf2v 15010                                                                                                                         1 ↵ ✹ ✭
Forwarding from [::1]:15010 -> 15010
Handling connection for 15010
```

```bash
go run pilot_cli.go -pilot "[::1]:15010" --proxytag sleep-9454cc476-pfmwj
```

and repeated the same, but this time for IPv4 only forwarding.

**Review by howardjohn — APPROVED** [review]

File: releasenotes/notes/44017.yaml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+apiVersion: release-notes/v2
+kind: feature
+area: traffic-management
+issue:
+  - 44017
+docs: []
+
+releaseNotes:
+- |
+  **Added** the user to specify the `IPFamilyPolicy` and `ipFamilies` setting in Istio Service resources either via the operator API or the helm charts.
```
Comment (hanxiaop) [inline releasenotes/notes/44017.yaml:10]:
```suggestion
  **Added** the ability for the user to specify the `IPFamilyPolicy` and `ipFamilies` setting in Istio Service resources either via the operator API or the helm charts.
```

**Review by hanxiaop — APPROVED** [review]
LGTM after improve the clarity of the release note.

**Pull request merged**